src/util/expr.h

```diff
@@ -161,6 +162,126 @@ class exprt:public irept
 
   void visit(class expr_visitort &visitor);
   void visit(class const_expr_visitort &visitor) const;
+
+  // Depth first search iterator - iterates over supplied expression
+  // and all its operands recursively.
+  template<typename expr_t>
+  class depth_iteratort final:
+  public std::iterator<std::forward_iterator_tag, expr_t, void>
+  {
+  public:
+    // Create end iterator
+    depth_iteratort() { }
+    // Create begin iterator, starting at the supplied node
+    explicit depth_iteratort(expr_t &root)
+    {
+      static_assert(std::is_same<exprt,
+      typename std::remove_const<expr_t>::type>::value,
+      "Only exprt and const exprt are allowed");
+      this->push_expr(root);
+    }
+    bool operator==(const depth_iteratort &other) const
+    {
+      return m_stack.size()==other.m_stack.size()
+            && std::equal(m_stack.begin(),
+                          m_stack.end(),
+                          other.m_stack.begin());
+    }
+    bool operator!=(const depth_iteratort &other) const
+    {
+      return !(*this==other);
+    }
+    depth_iteratort &operator++()
+    {
+      for(;;)
```

Comment:
Sorry, my example was too brief: if a depth_iterator is initialized to end() and that depth_iterator subsequently has `operator++` applied, then the access of `back()` is undefined.